projects/MAE/modeling/cross_entropy.py

```diff
@@ -0,0 +1,33 @@
+# coding=utf-8
+# Copyright 2021 The OneFlow Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import oneflow as flow
+import oneflow.nn as nn
+
+
+class SoftTargetCrossEntropy(nn.Module):
+    def __init__(self):
+        super(SoftTargetCrossEntropy, self).__init__()
+
+    def forward(self, x: flow.Tensor, target: flow.Tensor) -> flow.Tensor:
+        pred = flow.log_softmax(x, dim=-1)
+        loss = -target * pred
+        # sum and mean should be calculated with float32
+        # amp_white_identity ensure -target * pred using float16
+        # amp_black_identity ensure sum and mean using float32
+        loss = flow._C.amp_white_identity(loss)
+        loss = flow._C.amp_black_identity(loss)
```

Comment:
为了在这里控制后面 sum 和 mean 都用 float32 计算

Reply:
这里这个 amp__xx__identity 接口的设计，感觉有点让人迷惑，两个连着调用
感觉也是得知道实现细节的才知道怎么用，这个是不打算给用户用的是吧？

Reply:
这个一直没改，可以现在把这个接口改了，改成和算法无关的

Reply:
要不支持一下autocast的用法，只是graph和eager下autocast的行为不一样，lazy下autocast只是标记某个op是否可以自动cast

Reply:
没太理解houjiang说的：

- graph和eager下autocast的行为不一样：这里有什么不一样，以及为什么要实现成不一样的呢？
- lazy下autocast只是标记某个op是否可以自动cast：这里能解决这里 trick 的问题吗？

我理解这里 trick 的本质是：lazy amp 没有推导出我们理想的精度设置，所以需要人工干预，而人工干预又要顾虑 lazy amp 的特性：它会往前搜，它不会尊重 flow.cast 等用户意志。

lazy amp 为什么没有推导出我们理想的精度设置呢？因为其只考虑前驱后继的影响，但在深度学习实际业务中，后向也会受前向的精度影响。算法设计上没把这个业务知识考虑进去。比如前向中的 reduce sum 我们一般认为其是 black，但 broadcast add 展开的后向 reduce sum 一般认为其是 white。不能光考虑黑白名单。

Reply:
lazy amp 算法拿易用性换一点点潜在的自动优化可能，我个人觉得是不划算的。如果真有潜在优化的可能，不依靠自动，靠手动达到也行，这里的自动意义有多大，我是比较怀疑的。

Reply:
那更加合适的解决方案是什么呢？ @leaves-zwx  😂 

#### 原版：

```python
def forward(self, x: flow.Tensor, target: flow.Tensor) -> flow.Tensor:
        pred = flow.log_softmax(x, dim=-1)
        loss = -target * pred
        # sum and mean should be calculated with float32
        # amp_white_identity ensure -target * pred using float16
        # amp_black_identity ensure sum and mean using float32
        loss = flow._C.amp_white_identity(loss)  # 这两行出现在代码里，用户是看不懂的。至少要改一个接口
        loss = flow._C.amp_black_identity(loss)
        loss = flow.sum(loss, dim=-1)
        return loss.mean()
```


#### 改进版讨论

> 要不支持一下autocast的用法，只是graph和eager下autocast的行为不一样，lazy下autocast只是标记某个op是否可以自动cast

后江说的这个方案，是不是用这种方式来代替：

```python
def forward(self, x: flow.Tensor, target: flow.Tensor) -> flow.Tensor:
        pred = flow.log_softmax(x, dim=-1)
        loss = -target * pred
        with autocast(enabled=False): # 语义上是关闭 amp
            loss = flow.sum(loss.float(), dim=-1) # 手动让 sum 输入是 fp32
            loss = loss.mean()
        return loss
```

参考 pytorch 链接： https://pytorch.org/docs/stable/amp.html#torch.autocast

以及 pytorch 官网示例代码：

```python
# Creates some tensors in default dtype (here assumed to be float32)
a_float32 = torch.rand((8, 8), device="cuda")
b_float32 = torch.rand((8, 8), device="cuda")
c_float32 = torch.rand((8, 8), device="cuda")
d_float32 = torch.rand((8, 8), device="cuda")

with autocast():
    e_float16 = torch.mm(a_float32, b_float32)
    with autocast(enabled=False):
        # Calls e_float16.float() to ensure float32 execution
        # (necessary because e_float16 was created in an autocasted region)
        f_float32 = torch.mm(c_float32, e_float16.float())

    # No manual casts are required when re-entering the autocast-enabled region.
    # torch.mm again runs in float16 and produces float16 output, regardless of input types.
    g_float16 = torch.mm(d_float32, f_float32)
```